File: market-data/crypto-news/examples/python/feeds.py
#!/usr/bin/env python3
"""
Feeds & Export API Examples - Python
Free Crypto News API - https://github.com/nirholas/free-crypto-news

Examples for RSS feeds, exports, and data format endpoints.
"""

import requests
import json
from typing import Optional, List

BASE_URL = "https://cryptocurrency.cv"


# =============================================================================
# GET /api/rss - RSS Feed (XML)
# =============================================================================

def get_rss_feed(category: Optional[str] = None, limit: int = 50) -> str:
    """
    Get RSS feed in XML format.
    
    Args:
        category: Filter by category
        limit: Number of items
    
    Returns:
        RSS XML feed
    """
    params = {"limit": limit}
    if category:
        params["category"] = category
    
    response = requests.get(f"{BASE_URL}/api/rss", params=params)
    return response.text


# =============================================================================
# GET /api/rss.json - RSS Feed (JSON)
# =============================================================================

def get_rss_json(category: Optional[str] = None, limit: int = 50) -> dict:
    """
    Get RSS feed in JSON format.
    
    Args:
        category: Filter by category
        limit: Number of items
    
    Returns:
        RSS JSON feed
    """
    params = {"limit": limit}
    if category:
        params["category"] = category
    
    response = requests.get(f"{BASE_URL}/api/rss.json", params=params)
    return response.json()


# =============================================================================
# GET /api/feed/atom - Atom Feed
# =============================================================================

def get_atom_feed(category: Optional[str] = None, limit: int = 50) -> str:
    """
    Get Atom feed.
    
    Args:
        category: Filter by category
        limit: Number of items
    
    Returns:
        Atom XML feed
    """
    params = {"limit": limit}
    if category:
        params["category"] = category
    
    response = requests.get(f"{BASE_URL}/api/feed/atom", params=params)
    return response.text


# =============================================================================
# GET /api/feed/json - JSON Feed
# =============================================================================

def get_json_feed(category: Optional[str] = None, limit: int = 50) -> dict:
    """
    Get JSON Feed format.
    
    Args:
        category: Filter by category
        limit: Number of items
    
    Returns:
        JSON Feed
    """
    params = {"limit": limit}
    if category:
        params["category"] = category
    
    response = requests.get(f"{BASE_URL}/api/feed/json", params=params)
    return response.json()


# =============================================================================
# GET /api/export - Export Data
# =============================================================================

def export_data(format: str = "json", start: Optional[str] = None,
                end: Optional[str] = None, category: Optional[str] = None) -> dict:
    """
    Export news data in various formats.
    
    Args:
        format: Export format (json, csv, xml, parquet)
        start: Start date (YYYY-MM-DD)
        end: End date (YYYY-MM-DD)
        category: Filter by category
    
    Returns:
        Exported data or download URL
    """
    params = {"format": format}
    if start:
        params["start"] = start
    if end:
        params["end"] = end
    if category:
        params["category"] = category
    
    response = requests.get(f"{BASE_URL}/api/export", params=params)
    return response.json() if format == "json" else {"url": response.url}


# =============================================================================
# GET /api/export/csv - CSV Export
# =============================================================================

def export_csv(limit: int = 100, category: Optional[str] = None) -> str:
    """
    Export news data as CSV.
    
    Args:
        limit: Number of records
        category: Filter by category
    
    Returns:
        CSV data
    """
    params = {"limit": limit}
    if category:
        params["category"] = category
    
    response = requests.get(f"{BASE_URL}/api/export/csv", params=params)
    return response.text


# =============================================================================
# GET /api/export/json - JSON Export
# =============================================================================

def export_json(limit: int = 100, category: Optional[str] = None,
                pretty: bool = False) -> dict:
    """
    Export news data as JSON.
    
    Args:
        limit: Number of records
        category: Filter by category
        pretty: Pretty print JSON
    
    Returns:
        JSON data
    """
    params = {"limit": limit, "pretty": str(pretty).lower()}
    if category:
        params["category"] = category
    
    response = requests.get(f"{BASE_URL}/api/export/json", params=params)
    return response.json()


# =============================================================================
# GET /api/llms.txt - LLM-Friendly Format
# =============================================================================

def get_llms_txt() -> str:
    """
    Get news in LLM-friendly text format.
    
    Returns:
        Plain text formatted for LLMs
    """
    response = requests.get(f"{BASE_URL}/api/llms.txt")
    return response.text


# =============================================================================
# GET /api/sitemap - Sitemap
# =============================================================================

def get_sitemap() -> str:
    """
    Get XML sitemap.
    
    Returns:
        Sitemap XML
    """
    response = requests.get(f"{BASE_URL}/api/sitemap")
    return response.text


# =============================================================================
# GET /api/oembed - oEmbed
# =============================================================================

def get_oembed(url: str) -> dict:
    """
    Get oEmbed data for embedding.
    
    Args:
        url: Article URL
    
    Returns:
        oEmbed data
    """
    response = requests.get(f"{BASE_URL}/api/oembed", params={"url": url})
    return response.json()


# =============================================================================
# GET /api/embed - Embed Widget
# =============================================================================

def get_embed_code(theme: str = "light", limit: int = 5,
                   category: Optional[str] = None) -> dict:
    """
    Get embeddable widget code.
    
    Args:
        theme: Widget theme (light, dark)
        limit: Number of articles
        category: Filter by category
    
    Returns:
        Embed code and options
    """
    params = {"theme": theme, "limit": limit}
    if category:
        params["category"] = category
    
    response = requests.get(f"{BASE_URL}/api/embed", params=params)
    return response.json()


# =============================================================================
# GET /api/opml - OPML Export
# =============================================================================

def get_opml() -> str:
    """
    Get OPML file of all feeds.
    
    Returns:
        OPML XML
    """
    response = requests.get(f"{BASE_URL}/api/opml")
    return response.text


# =============================================================================
# GET /api/archive - News Archive
# =============================================================================

def get_archive(year: Optional[int] = None, month: Optional[int] = None,
                day: Optional[int] = None) -> dict:
    """
    Get archived news.
    
    Args:
        year: Year
        month: Month
        day: Day
    
    Returns:
        Archived articles
    """
    params = {}
    if year:
        params["year"] = year
    if month:
        params["month"] = month
    if day:
        params["day"] = day
    
    response = requests.get(f"{BASE_URL}/api/archive", params=params)
    return response.json()


# =============================================================================
# COMPLETE EXAMPLES
# =============================================================================

if __name__ == "__main__":
    print("\n" + "="*60)
    print("FREE CRYPTO NEWS API - FEEDS & EXPORT EXAMPLES")
    print("="*60)
    
    # 1. RSS Feed (XML)
    print("\n📡 1. RSS Feed (XML)")
    rss = get_rss_feed(limit=5)
    print(f"   RSS: {rss[:200]}...")
    
    # 2. RSS Feed (JSON)
    print("\n📊 2. RSS Feed (JSON)")
    rss_json = get_rss_json(limit=5)
    print(f"   JSON RSS: {rss_json}")
    
    # 3. Atom Feed
    print("\n⚛️ 3. Atom Feed")
    atom = get_atom_feed(limit=5)
    print(f"   Atom: {atom[:200]}...")
    
    # 4. JSON Feed
    print("\n📋 4. JSON Feed")
    json_feed = get_json_feed(limit=5)
    print(f"   JSON Feed: {json_feed}")
    
    # 5. CSV Export
    print("\n📑 5. CSV Export")
    csv = export_csv(limit=10)
    lines = csv.split('\n')[:3]
    print(f"   CSV Preview:\n   {lines}")
    
    # 6. JSON Export
    print("\n💾 6. JSON Export")
    json_export = export_json(limit=5, pretty=True)
    print(f"   JSON Export: {len(json_export)} items")
    
    # 7. LLMs.txt
    print("\n🤖 7. LLMs.txt Format")
    llms = get_llms_txt()
    print(f"   LLMs.txt: {llms[:200]}...")
    
    # 8. Sitemap
    print("\n🗺️ 8. Sitemap")
    sitemap = get_sitemap()
    print(f"   Sitemap: {sitemap[:200]}...")
    
    # 9. oEmbed
    print("\n🔗 9. oEmbed")
    oembed = get_oembed("https://example.com/article")
    print(f"   oEmbed: {oembed}")
    
    # 10. Embed Widget
    print("\n📦 10. Embed Widget")
    embed = get_embed_code(theme="dark", limit=5)
    print(f"   Embed: {embed}")
    
    # 11. OPML
    print("\n📂 11. OPML Export")
    opml = get_opml()
    print(f"   OPML: {opml[:200]}...")
    
    # 12. Archive
    print("\n📚 12. News Archive (2024)")
    archive = get_archive(year=2024, month=6)
    print(f"   Archive: {archive}")
    
    print("\n" + "="*60)
    print("Feeds & export examples completed!")
    print("="*60)


# =============================================================================
# INTEGRATION EXAMPLES
# =============================================================================

def subscribe_to_rss_in_reader():
    """Example: How to use RSS feed in a feed reader."""
    # Add this URL to your RSS reader:
    rss_url = f"{BASE_URL}/api/rss"
    
    # For specific categories:
    bitcoin_rss = f"{BASE_URL}/api/rss?category=bitcoin"
    defi_rss = f"{BASE_URL}/api/rss?category=defi"
    
    print(f"RSS Feed URLs:")
    print(f"  All News: {rss_url}")
    print(f"  Bitcoin: {bitcoin_rss}")
    print(f"  DeFi: {defi_rss}")


def import_to_pandas():
    """Example: Import data into pandas DataFrame."""
    try:
        import pandas as pd
        from io import StringIO
        
        csv_data = export_csv(limit=100)
        df = pd.read_csv(StringIO(csv_data))
        
        print(f"Loaded {len(df)} articles into DataFrame")
        print(f"Columns: {list(df.columns)}")
        return df
    except ImportError:
        print("pandas not installed. Run: pip install pandas")


def integrate_with_feedly():
    """Example: URLs to add to Feedly."""
    feeds = {
        "All Crypto News": f"{BASE_URL}/api/rss",
        "Bitcoin": f"{BASE_URL}/api/rss?category=bitcoin",
        "Ethereum": f"{BASE_URL}/api/rss?category=ethereum",
        "DeFi": f"{BASE_URL}/api/rss?category=defi",
        "NFT": f"{BASE_URL}/api/rss?category=nft",
        "Regulatory": f"{BASE_URL}/api/rss?category=regulatory",
    }
    
    print("Add these feeds to Feedly:")
    for name, url in feeds.items():
        print(f"  {name}: {url}")

File: market-data/crypto-news/examples/javascript/feeds.js
/**
 * Free Crypto News API - Feeds & Export Examples
 * https://github.com/nirholas/free-crypto-news
 * 
 * RSS, Atom, JSON feeds, exports, and integration formats.
 */

const BASE_URL = 'https://cryptocurrency.cv';

// =============================================================================
// RSS FEEDS
// =============================================================================

/**
 * Get RSS feed (XML format)
 * @param {Object} options - Query options
 * @returns {Promise<string>} RSS XML
 */
async function getRSSFeed(options = {}) {
  const params = new URLSearchParams({
    category: options.category || '',
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/rss?${params}`);
  return response.text();
}

/**
 * Get RSS feed as JSON
 * @param {Object} options - Query options
 * @returns {Promise<Object>} RSS as JSON
 */
async function getRSSJSON(options = {}) {
  const params = new URLSearchParams({
    category: options.category || '',
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/rss/json?${params}`);
  return response.json();
}

/**
 * Get OPML file for feed readers
 * @returns {Promise<string>} OPML XML
 */
async function getOPML() {
  const response = await fetch(`${BASE_URL}/api/feeds/opml`);
  return response.text();
}

// =============================================================================
// ATOM & JSON FEEDS
// =============================================================================

/**
 * Get Atom feed
 * @param {Object} options - Query options
 * @returns {Promise<string>} Atom XML
 */
async function getAtomFeed(options = {}) {
  const params = new URLSearchParams({
    category: options.category || '',
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/feeds/atom?${params}`);
  return response.text();
}

/**
 * Get JSON Feed (jsonfeed.org format)
 * @param {Object} options - Query options
 * @returns {Promise<Object>} JSON Feed
 */
async function getJSONFeed(options = {}) {
  const params = new URLSearchParams({
    category: options.category || '',
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/feeds/json?${params}`);
  return response.json();
}

// =============================================================================
// EXPORTS
// =============================================================================

/**
 * Export data in specified format
 * @param {string} format - Export format (csv, json, xlsx)
 * @param {Object} options - Query options
 * @returns {Promise<string|Object>} Exported data
 */
async function exportData(format, options = {}) {
  const params = new URLSearchParams({
    format,
    category: options.category || '',
    from: options.from || '',
    to: options.to || '',
    limit: options.limit || 100,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/export?${params}`);
  
  if (format === 'json') {
    return response.json();
  }
  return response.text();
}

/**
 * Export to CSV
 * @param {Object} options - Query options
 * @returns {Promise<string>} CSV data
 */
async function exportCSV(options = {}) {
  return exportData('csv', options);
}

/**
 * Export to JSON
 * @param {Object} options - Query options
 * @returns {Promise<Object>} JSON data
 */
async function exportJSON(options = {}) {
  return exportData('json', options);
}

// =============================================================================
// LLM & AI INTEGRATIONS
// =============================================================================

/**
 * Get llms.txt file (AI-readable format)
 * @returns {Promise<string>} llms.txt content
 */
async function getLLMsTxt() {
  const response = await fetch(`${BASE_URL}/llms.txt`);
  return response.text();
}

/**
 * Get llms-full.txt with complete context
 * @returns {Promise<string>} llms-full.txt content
 */
async function getLLMsFullTxt() {
  const response = await fetch(`${BASE_URL}/llms-full.txt`);
  return response.text();
}

// =============================================================================
// SITEMAP & SEO
// =============================================================================

/**
 * Get sitemap XML
 * @returns {Promise<string>} Sitemap XML
 */
async function getSitemap() {
  const response = await fetch(`${BASE_URL}/sitemap.xml`);
  return response.text();
}

/**
 * Get sitemap index
 * @returns {Promise<string>} Sitemap index XML
 */
async function getSitemapIndex() {
  const response = await fetch(`${BASE_URL}/sitemap-index.xml`);
  return response.text();
}

// =============================================================================
// EMBED & OEMBED
// =============================================================================

/**
 * Get oEmbed data for an article
 * @param {string} url - Article URL
 * @returns {Promise<Object>} oEmbed data
 */
async function getOEmbed(url) {
  const response = await fetch(`${BASE_URL}/api/oembed?url=${encodeURIComponent(url)}`);
  return response.json();
}

/**
 * Get embed code for an article
 * @param {string} articleId - Article ID
 * @param {Object} options - Embed options
 * @returns {Promise<Object>} Embed code
 */
async function getEmbedCode(articleId, options = {}) {
  const params = new URLSearchParams({
    id: articleId,
    theme: options.theme || 'light',
    width: options.width || '100%',
    height: options.height || 'auto',
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/embed?${params}`);
  return response.json();
}

// =============================================================================
// ARCHIVE
// =============================================================================

/**
 * Get archive for a specific date
 * @param {string} date - Date (YYYY-MM-DD)
 * @returns {Promise<Object>} Archive data
 */
async function getArchive(date) {
  const response = await fetch(`${BASE_URL}/api/archive?date=${date}`);
  return response.json();
}

/**
 * Get archive index
 * @returns {Promise<Object>} Archive index
 */
async function getArchiveIndex() {
  const response = await fetch(`${BASE_URL}/api/archive/index`);
  return response.json();
}

// =============================================================================
// WEBHOOKS
// =============================================================================

/**
 * Create webhook subscription
 * @param {Object} webhook - Webhook configuration
 * @returns {Promise<Object>} Webhook response
 */
async function createWebhook(webhook) {
  const response = await fetch(`${BASE_URL}/api/webhooks`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(webhook)
  });
  return response.json();
}

/**
 * List webhooks
 * @param {string} apiKey - API key
 * @returns {Promise<Object>} Webhooks list
 */
async function listWebhooks(apiKey) {
  const response = await fetch(`${BASE_URL}/api/webhooks`, {
    headers: { 'X-API-Key': apiKey }
  });
  return response.json();
}

/**
 * Delete webhook
 * @param {string} webhookId - Webhook ID
 * @param {string} apiKey - API key
 * @returns {Promise<Object>} Delete response
 */
async function deleteWebhook(webhookId, apiKey) {
  const response = await fetch(`${BASE_URL}/api/webhooks/${webhookId}`, {
    method: 'DELETE',
    headers: { 'X-API-Key': apiKey }
  });
  return response.json();
}

// =============================================================================
// EXPORTS
// =============================================================================

module.exports = {
  // RSS
  getRSSFeed,
  getRSSJSON,
  getOPML,
  
  // Atom & JSON Feed
  getAtomFeed,
  getJSONFeed,
  
  // Exports
  exportData,
  exportCSV,
  exportJSON,
  
  // LLM
  getLLMsTxt,
  getLLMsFullTxt,
  
  // Sitemap
  getSitemap,
  getSitemapIndex,
  
  // Embed
  getOEmbed,
  getEmbedCode,
  
  // Archive
  getArchive,
  getArchiveIndex,
  
  // Webhooks
  createWebhook,
  listWebhooks,
  deleteWebhook,
};

// =============================================================================
// EXAMPLE USAGE
// =============================================================================

async function runExamples() {
  console.log('\n=== Feeds & Export API Examples ===\n');
  
  try {
    // RSS JSON
    console.log('1. Fetching RSS feed as JSON...');
    const rss = await getRSSJSON({ limit: 5 });
    console.log(`   Items: ${rss.items?.length || 0}`);
    
    // JSON Feed
    console.log('2. Fetching JSON Feed...');
    const jsonFeed = await getJSONFeed({ limit: 5 });
    console.log(`   Items: ${jsonFeed.items?.length || 0}`);
    
    // Export JSON
    console.log('3. Exporting data as JSON...');
    const exported = await exportJSON({ limit: 5 });
    console.log(`   Exported: ${exported.data?.length || 0} items`);
    
    // LLMs.txt
    console.log('4. Fetching llms.txt...');
    const llms = await getLLMsTxt();
    console.log(`   Length: ${llms.length} characters`);
    
    // Archive index
    console.log('5. Fetching archive index...');
    const archiveIndex = await getArchiveIndex();
    console.log(`   Archives: ${archiveIndex.dates?.length || 0}`);
    
    // oEmbed
    console.log('6. Fetching oEmbed data...');
    const oembed = await getOEmbed('https://cryptocurrency.cv/article/example');
    console.log(`   Type: ${oembed.type || 'N/A'}`);
    
    console.log('\n✅ All feeds & export examples completed!\n');
  } catch (error) {
    console.error('Error:', error.message);
  }
}

// Run if executed directly
if (require.main === module) {
  runExamples();
}
